### ISSUE TITLE
Improve check messages

### ISSUE THREAD

**Issue by caugner** [checkstyle/checkstyle#3408]
Note: This is a follow-up of PR #3404 in which I improved the German translation (and for which I have thoroughly read all messages not only in German, but also in English).

**Context**:
Checkstyle displays warnings for occurrences in the checked code that don't fulfill the requirement of a check. These warnings are constructed from messages that are local to checks (and defined in [/src/main/resources/com/puppycrawl/tools/checkstyle/checks/](https://github.com/checkstyle/checkstyle/tree/master/src/main/resources/com/puppycrawl/tools/checkstyle/checks)). A message may have one or more parameters in order to better reflect the precise instance of a check violation.

**Problem**:
The messages vary in the degree to which they satisfy certain desirable quality criteria:

1. **Understandability** – A user should be able to understand what is wrong about his code with as little previous knowledge as possible.
2. **Guidance** – A user should be able to understand what he needs to do in order to fix his code with as little further knowledge as possible.
3. **Accuracy** – The message should accurately reflect what the check verifies.

In general, it is also noteworthy to check when to use "must" vs "should" and "must not" vs "should not".

**Solution**:
Messages should be improved where necessary and possible. An improvement can refer to:

* Making a message **more concise** by removing superfluous information.
* Making a message **more complete** by giving hints about background, consequence and/or fix.
* Making a message **easier to understand** by framing it in a sentence.
* Making a message **more accurate** by synchronizing with what the check actually/currently does.
* Making a message **more specific** by introducing additional message parameters.

**Examples**:
To give you an idea of what I'm talking about, here are some examples of messages that I reckon could be enhanced in order to significantly improve usability:

```properties
# There could be one message per style.
annotation.incorrect.style=Annotation style must be ''{0}''.

# This means: "The last value of an annotation array element should not have a trailing comma."
annotation.trailing.comma.present=Annotation array values cannot contain trailing comma.

# This means: "Always use @Foo() instead of @Foo for annotations".
annotation.parens.missing=Annotation must have closing parenthesis.

# This means: "If you implement a custom #equals method, you should always override #equals(Object) [to avoid confusion]."
covariant.equals=covariant equals without overriding equals(java.lang.Object).

# This means: "If you compare a String with a String literal, always put the literal on the left side of the equals comparison."
equals.avoid.null=String literal expressions should be on the left side of an equals comparison.

# This means: "Always use the uppercase "L" when defining long values."
upperEll=Should use uppercase ''L''.

# This means: "(Class Fan-Out Complexity) Class {2} depends on {0} other classes (max allowed is {1})."
classFanOutComplexity=Class Fan-Out Complexity is {0} (max allowed is {1}).

# This means: "Method {2} has {0} Non Commenting Source Statements (NCSS) (max allowed is {1})."
ncss.method=NCSS for this method is {0} (max allowed is {1}).
...
```

**Proposal**:
If you agree that the checkstyle messages could bear some overhaul, I would volunteer and start to assemble a list of suggestions in a commit (e.g. in a PR or in a fork branch), so that people can comment on it, before we make any preliminary or final decision.

**Comment by romani** [comment]
I appreciate you desire to help us to make checkstyle better to use.
We definitely have some problems with clear messages.

On the same side we already have 7 languages translations and them need to be updated too.
Additional problems is that we need some sort of voting on each change of message in English to make sure that all people who care can approve new wording.

BTW:
you can already do this without waiting for changes in main library - checkstyle. Look at http://checkstyle.sourceforge.net/config.html#Custom_messages 
This will help you to test your wording in your team and than come to us and propose already adopted changes.

**Comment by caugner** [comment]
> On the same side we already have 7 languages translations and them need to be updated too.

I agree that it is desirable to keep the other languages up-to-date, but I would also say that in case of doubt having _clear_ English messages _without_ up-to-date translations should still be better than having _unclear_ English messages _with_ up-to-date translations!? How about we try and get in contact with the contributors who last modified the translations as soon as we reach a consensus on the English messages?

> Additional problems is that we need some sort of voting on each change of message in English to make sure that all people who care can approve new wording.

I agree that everyone who wants to participate in this process should be invited to do so.

Yet given that there are 222 messages to review, I wouldn't have too much hope that so many fellow contributors are willing to participate here. So I would prefer to postpone the voting until it becomes necessary.

Moreover I see some inherent issues with votings, because they tend to be biased, especially when they're about "change" vs "no change" and "your proposal" vs "my proposal".

> you can already do this without waiting for changes in main library - checkstyle.

Thanks, I wasn't aware of that.

Personally, I'm not a big fan of customizations, because it only fixes the symptoms, not the cause. Moreover (and luckily), I have only observed a fraction of the Check messages, so the outcome of this field data would be very limited.

So here are two possible ways of how we could proceed:

1. We find a way to crowd source this and first get feedback on how clear (understandable, helpful and accurate) each message is, before we start working on improvement, beginning with the least clear messages.
2. We come up with a finished evaluation and proposal "ourselves" and gather feedback from the source later.

With regard to the proposed messages, I don't think that everyone should just submit proposal to any messages, because the result most likely won't be consistent, which would again decrease clarity for the end-user.
